jupyterlab_server/listings_handler.py

```diff
@@ -0,0 +1,73 @@
+"""Tornado handlers for dynamic listings loading."""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import os
+import time
+import json
+import re
+
+from .server import APIHandler
+
+import requests
+
+
+LISTINGS_URL_SUFFIX='@jupyterlab/extensionmanager-extension/listings.json'
+
+
+def fetch_listings():
+    if len(ListingsHandler.blacklist_uris) > 0:
+        blacklist = []
+        for blacklist_uri in ListingsHandler.blacklist_uris:
+            print('Fetching blacklist from {}'.format(ListingsHandler.blacklist_uris))
+            r = requests.request('GET', blacklist_uri, **ListingsHandler.listings_request_opts)
+            j = json.loads(r.text)
+            for b in j['blacklist']:
+                blacklist.append(b)
+            ListingsHandler.blacklist = blacklist
+    if len(ListingsHandler.whitelist_uris) > 0:
+        whitelist = []
+        for whitelist_uri in ListingsHandler.whitelist_uris:
+            print('Fetching whitelist from {}'.format(ListingsHandler.whitelist_uris))
+            r = requests.request('GET', whitelist_uri, **ListingsHandler.listings_request_opts)
+            j = json.loads(r.text)
+            for w in j['whitelist']:
+                whitelist.append(w)
+        ListingsHandler.whitelist = whitelist
+    ListingsHandler.listings = json.dumps({
+        'blacklist_uris': list(ListingsHandler.blacklist_uris),
+        'whitelist_uris': list(ListingsHandler.whitelist_uris),
+        'blacklist': ListingsHandler.blacklist,
+        'whitelist': ListingsHandler.whitelist,
+    })
+
+
+class ListingsHandler(APIHandler):
+    """An handler that returns the listings specs."""
+
+    blacklist_uris = set()
+    whitelist_uris = set()
+    blacklist = []
+    whitelist = []
+    listings = ''
```

Comment:
Seems like this is a duplicate of the data in the attributes above? Maybe just remove it and dynamically create that dict on each get? It makes it a bit harder to follow what's being set.

Reply:
These are class level attributes. I can remove them but would need to write some guards in the various methods (init, fetch...) that access them. IMHO this would make the code less readable than having these fields present. As workaround I have added some comments for each of these fields. Would it work for you?

Reply:
Ah sorry just mean remove the `listing` attribute not all the class attributes.

Reply:
`listing` attribute has been removed. thx.